<a href="https://colab.research.google.com/github/UnostentatiousDeveloper/MIAR_VIU_AlgoritmosOptimizacion/blob/main/Seminario_AlgoritmosGD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center>  **Algoritmos de optimización - Seminario**<br></center>
##### <center> **Nombre y Apellidos:** Gabriel Díaz Ireland</center>
##### <center> **Url:** https://github.com/UnostentatiousDeveloper/Algoritmos_de_Optimizacion_MIAR_VIU/blob/main/Seminario/Seminario_AlgoritmosGD.ipynb <br></center>

<br>
<br>
<br>

### **Índice** <br>

1. Problema de las Sesiones de doblaje <br>
2. Problema de la organización de los horarios de partidos de La Liga<br>
3. Combinar cifras y operaciones




**Recordar: Cuando aparece el astersico (*) la respuesta de ese apartado es obligatoria.**

IMPORTANTE - QUITAR los (*) donde estén, que joiden el formato.




<br>
<br>


                                        


## <center> **1. Problema de las Sesiones de doblaje** <br></center>


**Enunciado:**<br> Se precisa coordinar el doblaje de una película. Los actores del doblaje deben coincidir en las
tomas en las que sus personajes aparecen juntos en las diferentes tomas. Los actores de
doblaje cobran todos la misma cantidad por cada día que deben desplazarse hasta el estudio de
grabación independientemente del número de tomas que se graben. No es posible grabar más
de 6 tomas por día. El objetivo es planificar las sesiones por día de manera que el gasto por los
servicios de los actores de doblaje sea el menor posible.

Número de actores: 10 <br>
Número de tomas : 30 <br>
Actores/Tomas : https://bit.ly/36D8IuK<br>
- 1 indica que el actor participa en la toma
- 0 en caso contrario







                                        

In [ ]:

import pandas as pd
data1 = pd.read_excel('C:/Users/díazirg/OneDrive - Brambles/Gabriel/Routed/JL NOTEBOOKS/3.VIU_AI_Master/3.5_Algoritmos_Optimización/Seminario/Datos problema doblaje(30 tomas, 10 actores).xlsx', engine='openpyxl')

print("El número de tomas totales son 30 (30 filas)")
print("El número de actores totales son 10 (10 columnas) ")



actors_per_toma = (data1.apply(lambda row: set(row.index[row == 1]), axis=1).to_dict())

actors_per_toma[0].discard("Toma")

data1.head(30)

El número de tomas totales son 30 (30 filas)
El número de actores totales son 10 (10 columnas) 


Toma  1  2  3  4  5  6  7  8  9  10  Total
0      1  1  1  1  1  1  0  0  0  0   0      5
1      2  0  0  1  1  1  0  0  0  0   0      3
2      3  0  1  0  0  1  0  1  0  0   0      3
3      4  1  1  0  0  0  0  1  1  0   0      4
4      5  0  1  0  1  0  0  0  1  0   0      3
5      6  1  1  0  1  1  0  0  0  0   0      4
6      7  1  1  0  1  1  0  0  0  0   0      4
7      8  1  1  0  0  0  1  0  0  0   0      3
8      9  1  1  0  1  0  0  0  0  0   0      3
9     10  1  1  0  0  0  1  0  0  1   0      4
10    11  1  1  1  0  1  0  0  1  0   0      5
11    12  1  1  1  1  0  1  0  0  0   0      5
12    13  1  0  0  1  1  0  0  0  0   0      3
13    14  1  0  1  0  0  1  0  0  0   0      3
14    15  1  1  0  0  0  0  1  0  0   0      3
15    16  0  0  0  1  0  0  0  0  0   1      2
16    17  1  0  1  0  0  0  0  0  0   0      2
17    18  0  0  1  0  0  1  0  0  0   0      2
18    19  1  0  1  0  0  0  0  0  0   0      2
19    20  1  0  1  1  1  0  0  0  0   0      4
20    21  0  0  0  0  0  1  0  1  0   0      2
21    22  1  1  1  1  0  0  0  0  0   0      4
22    23  1  0  1  0  0  0  0  0  0   0      2
23    24  0  0  1  0  0  1  0  0  0   0      2
24    25  1  1  0  1  0  0  0  0  0   1      4
25    26  1  0  1  0  1  0  0  0  1   0      4
26    27  0  0  0  1  1  0  0  0  0   0      2
27    28  1  0  0  1  0  0  0  0  0   0      2
28    29  1  0  0  0  1  1  0  0  0   0      3
29    30  1  0  0  1  0  0  0  0  0   0      2

### <center>**1.1.1. ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?**<br></center>







Si tenemos 30 tomas y 10 actores; Si quitamos la restricción (Máximo de 6 tomas por día) tenemos que tener en cuenta que podemos grabar todas las tomas en un día (30 tomas; 1 días)

La solución óptima en este caso, y sin necesidad de programar, es grabar 30 tomas en un día y pagar a los 10 actores una sola vez. Coste final = Coste de desplazamiento * número de actores.

Las posibilidad para optimizar el coste en consecuencia, es única. 30 tomas. 1 día.


### <center>**1.1.2. ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?**</center>

En este caso tenemos que tener en cuenta que máximo de tomas por día son 6.

Idealmente, querríamos aprovechar y grabar 6 tomas al día siempre, ya que esto implicará un menor número de traslados a priori. En consecuencia, tenemos que 6 tomas al día para 30 tomas totales, tenemos un espacio de permutación 5 días.

**1- Espacio de subconjunto 6 en total de tomas 30: Usaremos las fórmula de la combinatoria sin repetición C(n, r) = n! / (r! * (n - r)!) también conocida como coeficiente binomial**

**2- Espacio de permutación sin repetición (5 días) P(n) = n!**


Ejecutamos en python:


In [ ]:
import math

n = 30
r = 6

combinations = math.comb(n, r) ## Paquete que ejecuta la binomial
print(f"\033[1mNúmero de combinaciones posibles: {combinations}\033[0m")

combinations

days = 5
permutations = math.factorial(days)  ## Paqeute que ejecuta el factorial
print(f"\033[1mNúmero de formas de ordenar una combinación en 5 días diferentes: {permutations}\033[0m")

Totalcomb = permutations * combinations
print(f"\033[1mCombinaciones totales de planificación de días y combinación de tomas: {Totalcomb}\033[0m")


Número de combinaciones posibles: 593775
Número de formas de ordenar una combinación en 5 días diferentes: 120
Combinaciones totales de planificación de días y tomas: 71253000


### <center> **1.2.1.  ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.** </center>

La mejor estructura de datos dependerás del algoritmo que se va a utilizar. En nuestro caso, la resolución final se trata de un algoritmo genético (Apartado 1.5.1). En consecuencia, la estructura de tabla binaria que se ha proporcionado en formato .csv es apropiada, ya que nos indica con un 0 y un 1 la existencia de cierto actor en cierta toma o no. Con esto podemos crear nuestro individuos, que serán la colección de distribución de personas en 5 grupos para realizar 6 tomas cada vez.

En consecuencia, la estructura final por la que se mueve el problema es la siguiente:

{
    'combinacionxi': [
        {
            'diaji': [
                {
                    'tomawi': {
                        'personasji': valor
                    },
                    'tomaxw': {
                        'personasji': valor
                    },
                     ... Otros tomas ...
                },
                 ... Otros días ...
            ],
        },
         ... Otras combinaciones ...
    ],
    'otracombinacionxj': [
         ... Otras combinaciones ...
    ],
     ... Otras combinaciones ...
}

Siendo cada combinación xi el individuo de nuestro algoritmo genético.


#### <center> **1.3.1. ¿Cual es la función objetivo?** </center>

La función objetivo es aquella que comprueba y suma el valor de coste de los actores en base a si están presentes en ese día o no. El objetivo será el siguiente.

 $$\text{Minimizar: } \sum_{i \in A} \sum_{j \in D} x_{ij} \cdot \text{Costo}_{\text{diario}} \cdot \text{Actores\_por\_dia}[j][i]$$


- A es el conjunto de todos los actores.

- D es el conjunto de días de grabación.

- xij es una variable binaria que toma el valor 1 si el actor i está presente en el día j, y 0 en caso contrario.

- Costo diario es el costo diario (Constante en nuestor problema)

#### <center> **1.3.2.  ¿Es un problema de maximización o minimización?** </center>

Como bien observamos en el apartado anterior, se trara de un problema de minimización. Se trata de que xij salga como un 0 el mayor número de veces posibles, reduciendo la cantidad de veces que sumamos el coste de un día de un empleado.


#### <center> **1.4.1. Diseña un algoritmo para resolver el problema por fuerza bruta** </center>

El siguiente CODIGO FUNCIONA: PERO ES MUY MUY LENTO

In [ ]:
from itertools import combinations

def calculate_cost(num_tomas_per_day):
    return num_tomas_per_day * COST_PER_DAY

def find_optimal_schedule_bruteforce(num_actors, num_tomas, max_tomas_per_day, actors_per_toma):
    min_cost = float('inf')
    optimal_schedule = None

    # Generar todas las combinaciones posibles de tomas por día
    for r in range(1, min(max_tomas_per_day, num_tomas) + 1):
        for combination in combinations(range(num_tomas), r):
            # Verificar si los actores requeridos para esta combinación están presentes
            actors_required = set()
            for toma in combination:
                actors_required.update(actors_per_toma[toma])

            if len(actors_required) == num_actors:
                # Verificar los cuatro días siguientes sin repeticiones
                total_cost = calculate_cost(len(combination))
                days = [combination]
                used_tomas = set(combination)
                for i in range(1, 5):
                    next_combination = combinations(range(num_tomas), r + i)
                    for next_comb in next_combination:
                        if not any(toma in used_tomas for toma in next_comb):
                            next_actors_required = set()
                            for next_toma in next_comb:
                                next_actors_required.update(actors_per_toma[next_toma])

                            if len(next_actors_required) == num_actors:
                                total_cost += calculate_cost(len(next_comb))
                                days.append(next_comb)
                                used_tomas.update(next_comb)
                                break

                if total_cost < min_cost:
                    min_cost = total_cost
                    optimal_schedule = days

    return optimal_schedule, min_cost

# Datos de entrada (ejemplo)
NUM_ACTORS = 10
NUM_TOMAS = 30
MAX_TOMAS_PER_DAY = 15
COST_PER_DAY = 100  # Supongamos que el costo por día de cada actor es de 100

# Representación de las tomas con los actores requeridos (ejemplo)
actors_per_toma = actors_per_toma

optimal_schedule, min_cost = find_optimal_schedule_bruteforce(NUM_ACTORS, NUM_TOMAS, MAX_TOMAS_PER_DAY, actors_per_toma)

print("Planificación óptima de sesiones por día:")
for day, combination in enumerate(optimal_schedule, 1):
    print(f"Día {day}: {combination}")
print("Gasto mínimo total: $", min_cost)


KeyboardInterrupt: 

### <center> **1.4.2. Calcula la complejidad del algoritmo por fuerza bruta** </center>

La complejidad del algoritmo en este caso es extremadamente elevada, ya que el resultado del proceso es el siguiente.

1 - Cálculo de todas las combinaciones posibles en un día (En otro apartado calculabamos 593775)

2 - Búsqueda del mejor set de combinaciones. Esto se hace iterando combinaciones de 6 en 6 días. Es decir, cojemos la combinación 1 con 2 con 3 con 4 con 5 y vamos iterando hasta que no quede nada que revisar en 1.

  - Truco aquí -> Podemos filtrar las tomas incluidas en el primer elemento en base a los elementos que vamos iterando, así evitando hacer calculos de costes innecesarios. Podríamos hacer búsqueda exhaustiva pero con Branch and Bound.

3 - Cada vez encontramos una combinación de días óptima, la guardamos en una variables coste que iremos comparando cada vez.

Como podemos observar, hay que iterar para cada toma viendo por fuerza bruta todas las posibles combinaciones con el resto de tomas. **Complejidad computacional esperada es O((n!)^k), donde n es el número de tomas (en este caso, n = 30), y k es el número de grupos (en este caso, k = 5).**

El algoritmo, no termina de ejecutarse (Por lo menos en mi ordenador)



### <center> **1.5.1. Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta** </center>

Nuestro enfoque y resolución, será un algorimto genético. Otras opciones posibles como hemos descrito anteriormente, podría ser un algoritmo de backtracking con unas restricciones bien asentadas.

In [ ]:
import random
import numpy as np
from deap import base, creator, tools, algorithms

def calculate_cost(tomas_per_day):
    return len(tomas_per_day) * COST_PER_DAY

def evaluate(individual):
    cost = sum([calculate_cost(tomas_per_day) for tomas_per_day in individual])
    return cost,

def find_optimal_schedule(num_actors, num_tomas, max_tomas_per_day, num_days=4, population_size=100, generations=50):
    actors = list(range(1, num_actors + 1))
    tomas = list(range(num_tomas))

    # Representación del individuo: cada día es una lista de tomas
    creator.create("FitnessMin", base.Fitness, weights=(-1.0,))
    creator.create("Individual", list, fitness=creator.FitnessMin)

    # Crear caja de herramientas
    toolbox = base.Toolbox()
    toolbox.register("tomaset", random.sample, tomas, max_tomas_per_day)
    toolbox.register("individual", tools.initRepeat, creator.Individual, toolbox.tomaset, n=num_days)
    toolbox.register("population", tools.initRepeat, list, toolbox.individual)

    toolbox.register("evaluate", evaluate)
    toolbox.register("mate", tools.cxTwoPoint)
    toolbox.register("mutate", tools.mutShuffleIndexes, indpb=0.1)
    toolbox.register("select", tools.selTournament, tournsize=3)

    population = toolbox.population(n=population_size)
    hof = tools.HallOfFame(1)

    # Ejecutar el algoritmo genético
    algorithms.eaSimple(population, toolbox, cxpb=0.5, mutpb=0.2, ngen=generations, halloffame=hof)

    return hof[0], hof[0].fitness.values[0]

# Datos de entrada
NUM_ACTORS = 10
NUM_TOMAS = 30
MAX_TOMAS_PER_DAY = 6
COST_PER_DAY = 100  # Supongamos que el costo por día de cada actor es de 100

# Representación de las tomas con los actores requeridos
actors_per_toma = actors_per_toma

# Ejecutar la función para encontrar la planificación óptima de sesiones por día
optimal_schedule, min_cost = find_optimal_schedule(NUM_ACTORS, NUM_TOMAS, MAX_TOMAS_PER_DAY)

print("Planificación óptima de sesiones por día:")
print(optimal_schedule)
print("Gasto mínimo: $", min_cost)



gen	nevals
0  	100   
1  	57    
2  	58    
3  	67    
4  	65    
5  	67    
6  	57    
7  	52    
8  	61    
9  	50    
10 	67    
11 	56    
12 	58    
13 	66    
14 	58    
15 	62    
16 	66    
17 	51    
18 	61    
19 	67    
20 	56    
21 	60    
22 	59    
23 	57    
24 	57    
25 	48    
26 	58    
27 	62    
28 	54    
29 	56    
30 	53    
31 	54    
32 	68    
33 	65    
34 	55    
35 	58    
36 	38    
37 	58    
38 	49    
39 	49    
40 	64    
41 	47    
42 	54    
43 	55    
44 	67    
45 	72    
46 	65    
47 	60    
48 	56    
49 	60    
50 	64    
Planificación óptima de sesiones por día:
[[25, 16, 22, 9, 19, 26], [6, 5, 18, 1, 29, 25], [6, 21, 19, 24, 18, 3], [3, 4, 29, 18, 8, 19]]
Gasto mínimo: $ 2400.0


### <center> **1.5.2. Calcula la complejidad del algoritmo** </center>

Para calcular la complejidad del algoritmo tenemos que tener en cuenta los siguientes aspectos.

**Complejidad de Espacio:**

num_actors: O(num_actors)
num_tomas: O(num_tomas)
num_days: O(num_days)
population_size: O(population_size * num_days)
generations: O(population_size * num_days)


**Complejidad de Tiempo:**

calculate_cost(tomas_per_day): O(num_tomas)   ### Como podemos observar, calculate_cost() no tiene complejidad de espacio. Su complejidad de espacio es la que necesite en cada iteración. (Lo mismo se aplica para el resto)

evaluate(individual): O(num_days * num_tomas)
find_optimal_schedule: O(population_size * num_days * max_tomas_per_day)
Operaciones genéticas (mate, mutate, select): O(population_size * num_days)

Cálculo final de complejidad de tiempo: O(population_size * num_days * max_tomas_per_day) + O(population_size * num_days * max_tomas_per_day) + O(generations * population_size * num_days * max_tomas_per_day) + O(generations * population_size * num_days * max_tomas_per_day) + O(generations * num_days * max_tomas_per_day)


La complejidad final del algoritmo genético en tiempoo es (generations * population_size * num_days * max_tomas_per_day), donde generations es el número de generaciones, pupulation_size es el tamaño de la población, num_days es el número de días en el que agrupamos las tomas y max_tomas_per_day es el límite máximo de tomas por día.

### <center> **1.6.1. Como crees que es posible avanzar en el estudio del problema. Ten en cuenta variaciones del problema y/o variaciones al alza del tamaño** </center>

Como desarollador, si pudiera dedicar varios meses a este problema, estos son los siguientes pasos que creo que sería conveniente tener en cuenta (Tanto para este set de datos como para otros que también representen el mismo problema con las tomas y los viajes de los actores).

1 - Buscaría una relación inductiva del set de datos, para empezar iterando en nuestros algoritmos heurísticos directamente desde un punto ideal de comienzo. Algunas posibles ideas son:
  1.1 - Buscar un orden lógico ¿Existe la posibilidad de que las tomas que más actores requieren, debieran ser distribuidas acorde primero? ¿Merece la pena utilizar un pequeño bactracking / "branch and bound" para encontrar ese punto de partida? O quizá utilizar valores del set de datos a partir de los actores, realizando estadística de cada actor según el número de veces que aparece en tomas y ordenando primero los que más aparezcan en todos los días.

2 - Intentaría utilzar técnicas alternativas. ¿Quizá las técnicas de grafos observadas en clase como kruskal y prim podrían funcionar?

Una buena opción, podría ser utilizar una nubes de puntos para acercanos al problema, donde las tomas más cercanas sean las que más actores contengan del mismo tipo, dandole un peso a cada actor característico para poder identificar el peso finald de la toma y hacer esa agrupación de puntos. En este caso los grafos que comunican dichos puntos deberían ser dirigidos y conexos.


## <center> **2. Problema distribución de partidos de la Liga** <br></center>


**Enunciado:** <br>  Desde la La Liga de fútbol profesional se pretende organizar los horarios de los partidos de
liga de cada jornada. Se conocen algunos datos que nos deben llevar a diseñar un
algoritmo que realice la asignación de los partidos a los horarios de forma que maximice
la audiencia.

Los horarios disponibles se conocen a priori y son los siguientes:

**Viernes** 20:00 <br>
**Sábado** 12:00,16:00,18:00,20:00 <br>
**Domingo** 12:00,16:00,18:00,20:00 <br>
**Lunes** 20:00 <br>

En primer lugar, se clasifican los equipos en tres categorías según el número de
seguidores (que tiene relación directa con la audiencia).

**Hay 3 equipos en la categoría A, 11 equipos de categoría B y 6 equipos de categoría C.**

Se conoce estadísticamente la audiencia que genera cada partido según los equipos
que se enfrentan y en horario de sábado a las 20h.

**Categoría A   |   Categoría A  = 2 Millones**   <br>
**Categoría A   |   Categoría B  = 1,3 Millones**  <br>
**Categoría A   |   Categoría C  = 1 Millones** <br>
**Categoría B   |   Categoría B  = 0.9 Millones** <br>
**Categoría B   |   Categoría B  = 0.75 Millones** <br>
**Categoría C   |   Categoría C  = 0.47 Millones** <br>



Si el horario del partido no se realiza a las 20 horas del sábado se sabe que se reduce
según los coeficientes de la siguiente tabla

**Viernes**
**20h - 0.4**

**Sábado**
**12h - 0.55**
**16h - 0.7**
**18h - 0.8**
**20h -1**

**Domingo**
**12h - 0.45**
**16h - 0.75**
**18h - 0.85**
**20h - 1**

**Lunes**
**20h - 0.4**

Debemos asignar obligatoriamente siempre un partido el viernes y un partido el lunes


Es posible la coincidencia de horarios, pero en este
caso la audiencia de cada partido se verá afectada y
se estima que se reduce en porcentaje según la
siguiente tabla dependiendo del número de
coincidencias

**Si por ejemplo 2 partidos coinciden, se resta un 25% de audiencia a cada coincidencia, es decir, el número de visualizadores total suponiendo que coincidan dos partidos AA y AA en sábado 20:00 sería AA  0.75 + AA  0.75**

**Si por ejemplo 3 partidos coinciden, se resta un 45% de audiencia a cada coincidencia, es decir, el número de visualizadores total suponiendo que coincidan tres partidos AA y AA y AA en sábado 20:00 sería AA  0.55 + AA  0.55  AA  0.55**

**Si por ejemplo 4 partidos coinciden, se resta un 60% de audiencia a cada coincidencia, es decir, el número de visualizadores total suponiendo que coincidan cuatro partidos AA y AA y AA y AA en sábado 20:00 sería AA  0.40 + AA  0.40 + AA  0.40 + AA  0.40**

Cabe destacar que a partir de coincidencias de 3 partidos, no sale rentable bajo ningún concepto, por lo que no lo tendremos en cuenta. Comprobamos esto en apartados más adelante.

In [ ]:
##Definimos nuestros datos##

import pandas as pd
import numpy as np
import os

data2_audiencia = {'AA':2e6,'AB':1.3e6,'AC':1e6,'BB':0.9e6,'BC':0.75e6,'CC':0.47e6}
data2_coeficiente = [
    ['v20h', 0.4],
    ['s12h', 0.55],
    ['s16h', 0.7],
    ['s18h', 0.8],
    ['s20h', 1],
    ['d12h', 0.45],
    ['d16h', 0.75],
    ['d18h', 0.85],
    ['d20h', 1],
    ['l20h', 0.4]
]

data2_coeficiente = pd.DataFrame(data2_coeficiente, columns=('dia', 'cr'))



print("\033[1mCoeficientes de horario\033[0m\n", data2_coeficiente.head(),"\n\n\033[1mAudiencia\033[0m\n" , data2_audiencia )


Coeficientes de horario
     dia    cr
0  v20h  0.40
1  s12h  0.55
2  s16h  0.70
3  s18h  0.80
4  s20h  1.00 

Audiencia
 {'AA': 2000000.0, 'AB': 1300000.0, 'AC': 1000000.0, 'BB': 900000.0, 'BC': 750000.0, 'CC': 470000.0}


### <center>**2.1.1. ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?**<br></center>





Las restricciones del problema son:

- Hay que tener al menos un partido en cada horario disponible tanto el viernes como en lunes

Si quitamos esta restricción las posibilidades serán las mismas que con restricciones. Al movernos en un espacio de 20 equipos (Hay 3 equipos en la categoría A, 11 equipos de categoría B y 6 equipos de categoría C.) y tener un total de 10 espacios disponibles, podemos determinar que siempre habrá un partido en lunes y otro en viernes.

Escogiendo una muestra con la mejor combinación y la peor combinación, podemos observar si merece la pena mover un partido de un horario malo a un horario bueno pero coincidente.

AA y AA mismo horario = 4 millones * 1 * 0.75
AA y AA diferente horario (peor horario posible) = 2 millones * 0.45   + 2millones * 1.

OJO - Podemos comprobar que sale más rentable tener dos partidos en el mejor horario (sábado 20:00) que un partido en otro horario para casos específicos (Mirar resultado del prompt de jupyter)

En cambio, con tres partidos coincidentes, esto no se cumple (Mirar resultado del prompt de jupyter)

AA y AA y AA mismo horario = 4 millones * 1 * 0.55
AA y AA diferente horario (peor horario posible) = 2 millones * 0.40 + 2 millones * 0.40   + 2millones * 1.


**La posibilidades teniendo en cuenta que no hay restricciones son de un coeficiente binomial con espacio k = 2 y espacio n = 20. A su vez se unen las posibilidades de permutación de los 10 elementos (10 partidos) en los 10 espacios distintos. Siendo posible jugar 8 partidos en un mismo horario, es decir, podemos llegar a ordenar desde 3 elementos, hasta 10 elementos.**

Teniendo en cuenta que tenemos 10 partidos en una jornada determinada y hay 10 horarios disponibles, las permutaciones son 10! .
En cambio, si coincidiesen partidos en un mismo horario  (por ejemplo 2 partidos coinciden), habría que tener en cuenta que no son 10 partidos los que tenemos que permutar, sino que son 9. Una vez vista esta relación calculamos estas probabilidades en Python



In [ ]:
import math
print("Dos partidos coincidentes en mejor horario", 4*1*0.75)
print("Dos partidos coincidentes en distinto horario",2*0.45 + 2*1)
print("")
print("Tres partidos coincidentes en mejor horario", 6*1*0.55 )
print("Tres partidos coincidentes en distinto horario", 2*0.40 + 2*0.40 + 2*1)
print("")
print("A partir de los tres partidos coindicentes y con nuestro set de datos, no sale rentable")



print("Teniendo en cuenta estas relacones podemos determinar que tendremos según nuestro horario 3 horario en los que coincidan 2 partidos y el resto se reparten. Dejano libre el domingo por la mañana, el lunes y el viernes.")

Dos partidos coincidentes en mejor horario 3.0
Dos partidos coincidentes en distinto horario 2.9

Tres partidos coincidentes en mejor horario 3.3000000000000003
Tres partidos coincidentes en distinto horario 3.6

A partir de los tres partidos coindicentes y con nuestro set de datos, no sale rentable
Teniendo en cuenta estas relacones podemos determinar que tendremos según nuestro horario 3 horario en los que coincidan 2 partidos y el resto se reparten. Dejano libre el domingo por la mañana, el lunes y el viernes.


### <center>**2.1.2. ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?**</center>
El objetivo del problema es organizar una jornada de fútbol con 10 partidos entre 20 equipos para obtener la máxima audiencia. Se busca asignar los coeficientes de reducción disponibles a los partidos con mayor audiencia para maximizar la audiencia total. Se puede jugar un partido el viernes y otro el lunes, y se considera estadísticamente la audiencia generada por cada partido a las 20h del sábado. Si se juega en otro horario, se aplica un coeficiente de reducción.

Para resolver el problema, se aplica un algoritmo de ordenación para ordenar los partidos de mayor a menor audiencia y los coeficientes de mayor a menor. Luego, se asignan los coeficientes a los partidos según su audiencia para maximizar la audiencia total.

La complejidad del problema depende de la complejidad del algoritmo de ordenación utilizado, que en el peor caso será O(n²).


### <center> **2.2.1.  ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.** </center>

La mejor estrucura de datos será la que contenga en una lista cada equipo con su respectiva categoría. Asi mismo, otra lista nos proporciona la audiencia generada en los partidos y otra list alos coeficientes disponibles.

In [ ]:
import pandas as pd

equipos = pd.DataFrame(columns=('nombre', 'categoria'))

equipos.loc[len(equipos)] = ['Equipo1', 'A']
equipos.loc[len(equipos)] = ['Equipo2', 'A']


equipos.loc[len(equipos)] = ['Equipo5', 'B']
equipos.loc[len(equipos)] = ['Equipo6', 'B']


equipos.loc[len(equipos)] = ['Equipo15', 'C']
equipos.loc[len(equipos)] = ['Equipo16', 'C']

equipos.head()


data2_audiencia = {'AA':2e6,'AB':1.3e6,'AC':1e6,'BB':0.9e6,'BC':0.75e6,'CC':0.47e6}
data2_coeficiente = [
    ['v20h', 0.4],
    ['s12h', 0.55],
    ['s16h', 0.7],
    ['s18h', 0.8],
    ['s20h', 1],
    ['d12h', 0.45],
    ['d16h', 0.75],
    ['d18h', 0.85],
    ['d20h', 1],
    ['l20h', 0.4]
]

print(equipos.head(),"\n\033[1mCoeficientes de horario\033[0m\n", data2_coeficiente,"\n\n\033[1mAudiencia\033[0m\n" , data2_audiencia )

     nombre categoria
0   Equipo1         A
1   Equipo2         A
2   Equipo5         B
3   Equipo6         B
4  Equipo15         C 
Coeficientes de horario
 [['v20h', 0.4], ['s12h', 0.55], ['s16h', 0.7], ['s18h', 0.8], ['s20h', 1], ['d12h', 0.45], ['d16h', 0.75], ['d18h', 0.85], ['d20h', 1], ['l20h', 0.4]] 

Audiencia
 {'AA': 2000000.0, 'AB': 1300000.0, 'AC': 1000000.0, 'BB': 900000.0, 'BC': 750000.0, 'CC': 470000.0}



### <center> **2.3.1. ¿Cual es la función objetivo?** </center>

La función objetivo viene definida por la siguiente relación.

**Audiencia en jornada:** Sumatorio(Audencia en jornada i, siendo i en rango de 1 a 10)

A su vez, para cada horario i, la fórmula es **audiencia_horario** = AudienciaPartidoj * coeficiente_coincidencia * coeficiente_horario

Juntando todo, la función objetivo sería:

**Audiencia_jornada = Σjornada j = (ΣAudienciaPartido_i1 * coeficiente_coincidencia_i1 * coeficiente_horario_i1 +  AudienciaPartido_i2 * coeficiente_coincidencia_i2 * coeficiente_horario_i2.... ) para i en posible rango de 1 a 10**


###<center> **2.3.2.  ¿Es un problema de maximización o minimización?** </center>


Es un problema de maximización, donde se intenta que el sumatorio de j sea el maxímo posible teniendo en cuanto la variación de coeficientes

### <center> **2.6.1. Como crees que es posible avanzar en el estudio del problema. Ten en cuenta variaciones del problema y/o variaciones al alza del tamaño** </center>

Se podrían tomar varias técnicas como resolución.

Restricciones previas y algoritmos lógicos: Con los cálculos que hemos demostrado, podemos ver cuando un coeficiente sale rentable o no. Podemos, antes de empezar de iterar y ordenar combinaciones, descartar automáticamente con una función, aquellos coeficientes que sabemos que no llevarána a la solución óptima. Quizá, con esta técnica previa, podamos dejar un espacio de posibilidades muy muy pequeño.

Algoritmos genéticos: Una muy buena opción para este caso, ya que podemos definir nuestras funciones de "fitness" en base a la función que queremos maximizar.

Algoritmos de búsqueda: Podría hacerse una prueba de conexión de grafos dirigidos. Se buscará siempre la función con mayor peso, sin poder repetir equipo para un conjunto de partidos dado. Esta solución podría ser interesante, ya que se parecería a fuerza bruta pero optimizando con alguna técnica de las revisadas en clase (Prim, Kruskal...)


## <center> **Problema 3. Combinar cifras y operaciones** <br></center>


**Enunciado:** El problema consiste en diseñar un algoritmo que lo resuelva.
• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)

Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:

4+2-6/3*1 = 4
• Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?
• Nota: Es posible usar la función de python “eval” para evaluar una expresión



Definimos los datos:

                                        

In [ ]:
import pandas as pd

nums = ['1','2','3','4','5','6','7','8','9']
signs  = ['/', '-', '+', '*']

print(nums, signs)

['1', '2', '3', '4', '5', '6', '7', '8', '9'] ['/', '-', '+', '*']


### <center>**3.0.0. Preguntas cartacterísticas del ejercicio 3**<br></center>

#### **¿Que valor máximo y mínimo se pueden obtener? OJO tomamos como referencia el enunciado, que dice números enteros (negativos incluidos)**

Mediante lógica, repartimos los números afines a obtener el máximo y el mínimo.

**Máximo: 7-1/5+8*9= 78.8** <br>
**Mínimo: 1+2/5-8*9 = -70.6**

Además, comprobamos esta solución en el apartado (3.3.1) por fuerza bruta.


#### **¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?**

Si es posible, en el apartado (3.3.1) realizamos la resolución. Podemso observar también que no todos los números racioanles tienen combinación posible.


### <center>**3.1.1. ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?**<br></center>





En este caso tenemos que tener en cuenta cuales son todas las restricciones.

Idealmente y quitando cualquier restricción, las posibilidades son infinitas dentro del conjunto de números Reales (Excluyendo los Irracionales). **P = {x ∈ R : x ∉ I}**. Esto es así por que podemos realizar cualquier combinación de símbolo con cualquier combinación de número.

Las nuevas normas son: podemos usar cualquier número el número de veces que y combinado con el símbolo que queramos las veces que queramos. Por ejemplo. Para calcular -1000 podríamos hacer 1-1-1-1-1-1..... repitiendo la combinación de caractéres "-1" 1001 veces.

### <center>**3.1.2. ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?**</center>

En este caso tenemos que tener en cuenta que HAY que combinar todos los signos posibles (+ ,- , / , * ) para llegar nuestro número objetivo.

En consecuencia, tenemos una lista de 9 números (1,2,3,4,5,6,7,8,9) y una lista de 4 caractéres (+ ,- , / , * ) y tenemos que usar esos 4 signos siempre.

En consecuencia, esta es nuestra fórmula de combinación de caractéres:

n = (c1,b1,c2,b2,c3,b3,c4,b4,c5) siendo (ci = signo)  y (bi = números).

Las posibles combinaciones son (y sin tener en cuenta repeticiones):


**Combinaciones de números:**

- c1,c2,c3,c4,c5 (k) ; 1,2,3,4,5,6,7,8,9 (n)

**Combinaciones de signos:**

- Permutación +, -, *, /

- Combinaciones totales:

- Combinaciones números * Combinaciones signos


In [ ]:
import math


n = 9
r = 5

combinations = math.perm(n, r) ## Paquete que ejecuta la binomial
print(f"\033[1mNúmero de combinaciones de 9 números en 5 espacios posibles: {combinations}\033[0m")



signsn = 4
permutations = math.factorial(signsn)  ## Paqeute que ejecuta el factorial
print(f"\033[1mNúmero de formas de ordenar signos en 4 espacios: {permutations}\033[0m")

Totalcomb = permutations * combinations
print(f"\033[1mCombinaciones totales de números con signos: {Totalcomb}\033[0m")
print(f"\033[1mCombinaciones totales de números con signos sin repetición de resultado: 242497\033[0m")

Número de combinaciones de 9 números en 5 espacios posibles: 15120
Número de formas de ordenar signos en 4 espacios: 24
Combinaciones totales de números con signos: 362880
Combinaciones totales de números con signos sin repetición de resultado: 242497


### <center> **3.2.1.  ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.** </center>

Este ejercicio se adapta muy bien a una estructura de combinación por caracteres, tomando una lista [1:9] y ciengo los números como tipo de dato "type string". Por otro lado, cogemos una lista de signos [+ ,- , / , *].

A partir de aquí, independientemente del algoritmo que usemos, podemos hacer combinaciones entre ambas listas. Por ejemplo para fuerza bruta realizando una doble iteración, o para un algoritmo genético, realizando un primer acercamiento del individuo a formarse en base a los caracteres disponibles (Se use o no otras técnicas inductivas previas).

Quizá, otra opción sería usar un acercamiento matemático (lógico) probado previamente, en cuyo caso, no haría falta esta estructura inicial, si no mediante programación dinámica ir guardando los caractéres utilizados previamente.


#### <center> **3.3.1. ¿Cual es la función objetivo?** </center>

La función objetivo es aquella que comprueba una combinación que nos da el número n que se pide.

n = (c1,b1,c2,b2,c3,b3,c4,b4,c5)

siendo (ci = signo)  y (bi = números)

Se puede revisar en el siguiente apartado el desarollo de la función implementada mediante fuerza bruta.

#### <center> **4.4.1. Diseña un algoritmo para resolver el problema por fuerza bruta** </center>

El siguiente CODIGO FUNCIONA: PERO ES MUY MUY LENTO. Probamos desde -100 a 100 irmero. Finalmente para obtener detalle decimal, probamos desde -80 a 81.

In [ ]:
import itertools
from itertools import combinations
import numpy as np

combn = []


signscomb= []

count = -1
for i in signs:
    signsav = [sign for sign in signs if sign != i]
    count = count + 1
    for j in range(0,3,1):
        signsav2 = [sign for sign in signsav if sign != signsav[j]]
        for v in range(0,2,1):
            signsav3 = [sign for sign in signsav2 if sign != signsav2[v]]
            comb = [signs[count] , signsav[j] , signsav2[v] , signsav3[0]]
            signscomb.append(comb)


print("All signs combinations ",len(signscomb), signscomb)


nums_comb = list(itertools.permutations(range(1, 10), 5))

# Print the number of combinations and the list of combinations
print(" All Number  combinations:", len(nums_comb),nums_comb)

def find_expres(p,it_count=0):
    for j in nums_comb:
        for i in signscomb:
            total_comb = [str(j[0])+i[0]+str(j[1])+i[1]+str(j[2])+i[2]+str(j[3])+i[3]+str(j[4])]
            it_count +=1
            if eval(total_comb[0]) == p:
                return print("Number of iterations: ", it_count, " Asked number: ", p ," Returned Expression: ", total_comb)
    return print("No existant expression for number asked: ", p, " Number of iterations:", it_count)


find_expres(2)
find_expres(7)
find_expres(234)
find_expres(77)

for i in [round(x, 1) for x in np.arange(-80, 81, 0.1)]:
    find_expres(i)


All signs combinations  24 [['/', '-', '+', '*'], ['/', '-', '*', '+'], ['/', '+', '-', '*'], ['/', '+', '*', '-'], ['/', '*', '-', '+'], ['/', '*', '+', '-'], ['-', '/', '+', '*'], ['-', '/', '*', '+'], ['-', '+', '/', '*'], ['-', '+', '*', '/'], ['-', '*', '/', '+'], ['-', '*', '+', '/'], ['+', '/', '-', '*'], ['+', '/', '*', '-'], ['+', '-', '/', '*'], ['+', '-', '*', '/'], ['+', '*', '/', '-'], ['+', '*', '-', '/'], ['*', '/', '-', '+'], ['*', '/', '+', '-'], ['*', '-', '/', '+'], ['*', '-', '+', '/'], ['*', '+', '/', '-'], ['*', '+', '-', '/']]
 All Number  combinations: 15120 [(1, 2, 3, 4, 5), (1, 2, 3, 4, 6), (1, 2, 3, 4, 7), (1, 2, 3, 4, 8), (1, 2, 3, 4, 9), (1, 2, 3, 5, 4), (1, 2, 3, 5, 6), (1, 2, 3, 5, 7), (1, 2, 3, 5, 8), (1, 2, 3, 5, 9), (1, 2, 3, 6, 4), (1, 2, 3, 6, 5), (1, 2, 3, 6, 7), (1, 2, 3, 6, 8), (1, 2, 3, 6, 9), (1, 2, 3, 7, 4), (1, 2, 3, 7, 5), (1, 2, 3, 7, 6), (1, 2, 3, 7, 8), (1, 2, 3, 7, 9), (1, 2, 3, 8, 4), (1, 2, 3, 8, 5), (1, 2, 3, 8, 6), (1, 2, 3, 8, 7), (1

### <center> **4.4.2. Calcula la complejidad del algoritmo por fuerza bruta** </center>

El algoritmo genera expresiones matemáticas para evaluar un número objetivo usando combinaciones de operadores y números. La complejidad en tiempo es alta debido a los bucles anidados que recorren todas las combinaciones posibles de operadores y números, lo que resulta en O(m! * n^2). La complejidad en espacio también es significativa debido a las grandes listas de permutaciones de números, siendo O(15,120), y las combinaciones de operadores, siendo O(1). En resumen, el algoritmo es ineficiente en términos de tiempo y espacio.

Si entramos un poco más en detalle en la complejidad de tiempo podemos concluir que:

Generación de combinaciones de operadores (signscomb):
Complejidad: O(n^2), donde "n" es el número de operadores.

Generación de permutaciones de números (nums_comb):
Complejidad: O(m!), donde "m" es el número de elementos en nums_comb.

Función find_expres:
Complejidad: O(m! * n^2), donde "m" es el número de permutaciones en nums_comb y "n" es el número de operadores.

Bucle final que invoca find_expres repetidamente para valores entre -80 y 81 teniendo en cuenta un decimal:
Complejidad: O((m! * n^2) * 160 * 10) con la prueba que hemos hecho nosotros.





### <center> **4.5.1. Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta** </center>

Nuestro enfoque y resolución, será un algorimto aleatorio. Otras opciones posibles como hemos descrito anteriormente, podría ser un algoritmo de lógica deductiva, siguiendo patrones lógicos matemáticos estabelcidos previamente. Otra opción serían algoritmos evolutivos, interesante de explorar con el tiempo necesario.

Como ya hemos usado algoritmos evolutivos en el ejercicio 1 (1.3.1) para este apartado vamos a utilizar un algoritmo aleatorio. ¿Tardará más en encontrar las soluciones sin un orden?¿O se tarda menos por fuerza bruta?

In [ ]:
import itertools
import numpy as np
import random

signs = ['+', '-', '*', '/']

signscomb = list(itertools.permutations(signs, 4))

nums_comb = list(itertools.permutations(range(1, 10), 5))
print(nums_comb)
print(nums_comb)
def generate_expression(nums, signs):
    # Generate a random expression using numbers and signs without repetition
    nums_sample = random.choice(nums)
    signs_sample = random.choice(signs)
    expression = [
        str(nums_sample[0]) + signs_sample[0] + str(nums_sample[1]) + signs_sample[1] +
        str(nums_sample[2]) + signs_sample[2] + str(nums_sample[3]) + signs_sample[3] +
        str(nums_sample[4])
    ]
    return expression[0]

def find_expres(p, it_count=0):
    for _ in range(1000):  # Run 1000 iterations to find the expression
        expression = generate_expression(nums_comb, signscomb)
        if eval(expression) == p:
            return print("Numero de iteraciones:", it_count + 1, "Número:", p, "Expresión:", expression)
        it_count += 1

    return print("No existe expresión, tras 1000 iteraciones:", p, "Numero de iteraciones:", it_count)

find_expres(2)
find_expres(7)
find_expres(234)
find_expres(76)

[(1, 2, 3, 4, 5), (1, 2, 3, 4, 6), (1, 2, 3, 4, 7), (1, 2, 3, 4, 8), (1, 2, 3, 4, 9), (1, 2, 3, 5, 4), (1, 2, 3, 5, 6), (1, 2, 3, 5, 7), (1, 2, 3, 5, 8), (1, 2, 3, 5, 9), (1, 2, 3, 6, 4), (1, 2, 3, 6, 5), (1, 2, 3, 6, 7), (1, 2, 3, 6, 8), (1, 2, 3, 6, 9), (1, 2, 3, 7, 4), (1, 2, 3, 7, 5), (1, 2, 3, 7, 6), (1, 2, 3, 7, 8), (1, 2, 3, 7, 9), (1, 2, 3, 8, 4), (1, 2, 3, 8, 5), (1, 2, 3, 8, 6), (1, 2, 3, 8, 7), (1, 2, 3, 8, 9), (1, 2, 3, 9, 4), (1, 2, 3, 9, 5), (1, 2, 3, 9, 6), (1, 2, 3, 9, 7), (1, 2, 3, 9, 8), (1, 2, 4, 3, 5), (1, 2, 4, 3, 6), (1, 2, 4, 3, 7), (1, 2, 4, 3, 8), (1, 2, 4, 3, 9), (1, 2, 4, 5, 3), (1, 2, 4, 5, 6), (1, 2, 4, 5, 7), (1, 2, 4, 5, 8), (1, 2, 4, 5, 9), (1, 2, 4, 6, 3), (1, 2, 4, 6, 5), (1, 2, 4, 6, 7), (1, 2, 4, 6, 8), (1, 2, 4, 6, 9), (1, 2, 4, 7, 3), (1, 2, 4, 7, 5), (1, 2, 4, 7, 6), (1, 2, 4, 7, 8), (1, 2, 4, 7, 9), (1, 2, 4, 8, 3), (1, 2, 4, 8, 5), (1, 2, 4, 8, 6), (1, 2, 4, 8, 7), (1, 2, 4, 8, 9), (1, 2, 4, 9, 3), (1, 2, 4, 9, 5), (1, 2, 4, 9, 6), (1, 2, 4, 9, 

Como podemos observar, el número de iteraciones baja muy notablemente. Una posible mejora de este algoritmo es ver que número o que rango de números no encuentran soluciones tras x iteraciones random y que rango si encuentra rápidamente. Consiguiendo combinar una lógica inductiva en la resolución, a partir de restricciones previas que mejoren el rendimiento. Números más dificiles de obtener mediante combinaciones (como puede ser 76) se verían beneficiados por esta lógica.

### <center> **4.5.2. Calcula la complejidad del nuevo algoritmo** </center>

Igual que el algoritmo de fuerza bruta, la complejidad se repite en algunos casos:


**Se repite: Generación de combinaciones de operadores (signscomb):**
**Complejidad: O(n^2), donde "n" es el número de operadores.**

**Se repite: Generación de permutaciones de números (nums_comb):**
**Complejidad: O(m!), donde "m" es el número de elementos en nums_comb.**

**Nuevo: random.sample es O(k)**

**Nuevo: La función find_expres ejecuta un bucle for con 1000 iteraciones. En cada iteración, llama a generate_expression, lo que implica la complejidad de O(k) como se mencionó anteriormente. Por lo tanto, la complejidad de find_expres es O(1000 * k), que se puede aproximar a O(k)**

Como hemos mencionado anteriormente, esta técnica será mucho peor cuanto más aumentemos el espacio. Es decir, si por ejemplo queremos incluir decimales, la eficiencia y complejidad de este algoritmo empeora. Y no siempre encontrará el resultado.

### <center> **4.6.1. Como crees que es posible avanzar en el estudio del problema. Ten en cuenta variaciones del problema y/o variaciones al alza del tamaño** </center>


Quizá, pudieramos mejorar el algoritmo actual, ajustando sus parámetros y optimizando su rendimiento. Un buen acercameinto (mencionado anterirormente) es la búsqueda exhaustiva que nos permite probar todas las combinaciones posibles de números y operadores, aunque puede ser computacionalmente costosa.

Otra opción (también mencioanda anteriormente) son los algoritmos genéticos. Podría estar interesante tratar las eppresiones matemáticas como individuos y regular su fitness y su evolución mediante aproximaciones. Aunque aquí, quizá habrái que evaluar la tendencia de la función como tal llendo de un lado hacia otro al cambiar la mutación.

Es decir, habría que "ayudar" al gen mutante con un descenso del gradiente de la función actual. Aunque esto solo es una idea.

### **SUCIIO y Anotaciones 1 -**

In [ ]:
from itertools import combinations

def calculate_cost(actors_per_day):
    return len(actors_per_day) * COST_PER_DAY

def find_optimal_schedule(num_actors, num_tomas, max_tomas_per_day):
    actors = list(range(1, num_actors + 1))
    tomas = list(range(num_tomas))
    cost = 0
    optimal_schedule = None
    min_cost = float('inf')

    # Generar todas las combinaciones posibles de tomas por día
    for r in range(1, min(max_tomas_per_day, num_tomas) + 1):


        for combination in combinations(tomas, r):

            # Verificar si los actores requeridos para esta combinación están presentes
            actors_required = set()
            for toma in combination:

                print(cost)
                round = 0
                round+=1
                print('round ' , round)


                actors_required.update(actors_per_toma[toma])

                print(actors_required)

            if len(actors_required) == num_actors:
                cost = calculate_cost(combination)

                if cost < min_cost:
                    min_cost = cost
                    optimal_schedule = combination

    return optimal_schedule, min_cost

# Datos de entrada
NUM_ACTORS = 10
NUM_TOMAS = 30
MAX_TOMAS_PER_DAY = 6
COST_PER_DAY = 100  # Supongamos que el costo por día de cada actor es de 100

# Representación de las tomas con los actores requeridos
actors_per_toma = actors_per_toma

optimal_schedule, min_cost = find_optimal_schedule(NUM_ACTORS, NUM_TOMAS, MAX_TOMAS_PER_DAY)

print("Planificación óptima de sesiones por día:")
print(optimal_schedule)
print("Gasto mínimo: $", min_cost)
